sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Throttle.java

```diff
@@ -0,0 +1,469 @@
+      return new OffsetRange(-1, size);
+    }
+
+    /**
+     * The {@link GetInitialWatermarkEstimatorState} initializes to this DoFn's output watermark to
+     * a negative infinity timestamp via {@link BoundedWindow#TIMESTAMP_MIN_VALUE}. The {@link
+     * Instant} returned by this method provides the runner the value is passes as an argument to
+     * this DoFn's {@link #newWatermarkEstimator}.
+     */
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkState() {
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * This DoFn uses a {@link WatermarkEstimators.Manual} as its {@link NewWatermarkEstimator},
+     * instantiated from an {@link Instant}. The state argument in this method comes from the return
+     * of the {@link #getInitialWatermarkState}.
+     */
+    @NewWatermarkEstimator
+    public WatermarkEstimator<Instant> newWatermarkEstimator(
+        @WatermarkEstimatorState Instant state) {
+      return new WatermarkEstimators.Manual(state);
+    }
+
+    /** Instantiates an {@link OffsetRangeTracker} from an {@link OffsetRange} instance. */
+    @NewTracker
+    public RestrictionTracker<OffsetRange, Long> newTracker(@Restriction OffsetRange restriction) {
+      return new OffsetRangeTracker(restriction);
+    }
+
+    /** Simply returns the {@link OffsetRange} restriction. */
+    @TruncateRestriction
+    public RestrictionTracker.TruncateResult<OffsetRange> truncate(
+        @Restriction OffsetRange restriction) {
+      return new RestrictionTracker.TruncateResult<OffsetRange>() {
+        @Override
+        public OffsetRange getTruncatedRestriction() {
+          return restriction;
+        }
+      };
+    }
+
+    /**
+     * Emits the next {@code element.getValue().get(position)} from the {@link
+     * OffsetRange#getFrom()} if {@link RestrictionTracker#tryClaim} is true and sets the watermark.
+     * If remaining items exist, it returns {@link
+     * DoFn.ProcessContinuation#withResumeDelay(Duration)} with {@link
+     * Configuration#getMaximumRate()}'s {@link Rate#getInterval()}, otherwise {@link
+     * ProcessContinuation#stop()}.
+     */
+    @ProcessElement
+    public ProcessContinuation process(
+        @Element KV<Integer, List<RequestT>> element,
+        @Timestamp Instant timestamp,
+        ManualWatermarkEstimator<Instant> estimator,
+        RestrictionTracker<OffsetRange, Long> tracker,
+        OutputReceiver<RequestT> receiver) {
+
+      if (element.getValue() == null || element.getValue().isEmpty()) {
+        return ProcessContinuation.stop();
+      }
+
+      long position = tracker.currentRestriction().getFrom();
+      if (position < 0) {
+        position = 0;
+      }
+
+      if (!tracker.tryClaim(position)) {
+        return ProcessContinuation.stop();
+      }
+
+      RequestT value = element.getValue().get((int) position);
+      estimator.setWatermark(timestamp);
+      receiver.output(value);
```

Comment:
I keep getting hung up on this function, even with the current approach (which I think has addressed many of my concerns). Basically, my remaining question is: in streaming mode does this actually guarantee the throttling behavior we're expecting. Lets say, for example, that we have a throttling limit of 100 elements per hour. And then the following happens:

```
We get 100 elements all at once
These elements are grouped into batches of size 1, then processed by this transform, all of them fire
maxBufferingDuration time passes (but less than an hour has passed, so we should still be throttled)
We get 100 elements all at once again
```
won't those 100 elements come in, get put into batches of size 1, and then immediately fire (since they have a new restriction tracker/watermark)?

--------------------------------------

FWIW - I do think this is the type of conversation that would have benefited greatly from being in a design doc where we can do threaded conversation more quickly. Not asking you to retrofit this in, but I am asking that we consider more detailed designs in the future to avoid getting bogged down like we have. Each time we've return to the PR, we've done so with very different approaches, which is expensive for you as an implementer and me as a reviewer.

Reply:
One possible approach is to have a stateful step using OrderedListState where you wrap each element with the desired firing time, then using an SDF kind of like this one to return a process continuation until we're ready to fire the element (this gets around the issues you've run into with timers not firing correctly in batch mode). This may still have watermark concerns though (mentioned elsewhere)

Reply:
> One possible approach is to have a stateful step using OrderedListState where you wrap each element with the desired firing time, then using an SDF kind of like this one to return a process continuation until we're ready to fire the element (this gets around the issues you've run into with timers not firing correctly in batch mode). This may still have watermark concerns though (mentioned elsewhere)

I really like this idea. However, [Timers and state must not be used.](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/DoFn.ProcessElement.html#:~:text=Timers%20and%20state%20must%20not%20be%20used.)

Reply:
You can use state in one transform then an SDF in the next. However, I still have concerns about watermarks and SDFs which need to be resolved before we would proceed with this approach (see my next comment)